src/model_theory/terms_and_formulas.lean

```diff
@@ -362,6 +379,137 @@ begin
       { exact (dif_neg (λ h', h (nat.lt_of_add_lt_add_left h'))).trans (dif_neg h).symm } } }
 end
 
+/-- An atomic formula is either equality or a relation symbol applied to terms. -/
+inductive is_atomic : L.bounded_formula α n → Prop
+| equal {t₁ t₂ : L.term (α ⊕ fin n)} : is_atomic (bd_equal t₁ t₂)
+| rel {l : ℕ} {R : L.relations l} {ts : fin l → L.term (α ⊕ fin n)} :
+    is_atomic (R.bounded_formula ts)
+
+lemma is_atomic.induction {P : L.bounded_formula α n → Prop} {φ : L.bounded_formula α n}
+  (h : is_atomic φ)
+  (he : ∀ (t₁ t₂ : L.term (α ⊕ fin n)), P (bd_equal t₁ t₂))
+  (hr : ∀ {l : ℕ} (R : L.relations l) (ts : fin l → L.term (α ⊕ fin n)), P (R.bounded_formula ts)) :
+  P φ :=
+begin
+  induction h,
+  { apply he },
+  { apply hr }
+end
+
+lemma is_atomic_bd_equal (t₁ t₂ : L.term (α ⊕ fin n)) :
+  is_atomic (t₁.bd_equal t₂ : L.bounded_formula α n) :=
+is_atomic.equal
+
+lemma is_atomic_rel {m} (R : L.relations m) (ts) :
+  is_atomic (R.bounded_formula ts : L.bounded_formula α n) :=
+is_atomic.rel
+
+lemma is_atomic.relabel {m : ℕ} {φ : L.bounded_formula α m} (h : φ.is_atomic)
+  (f : α → β ⊕ (fin n)) :
+  (φ.relabel f).is_atomic :=
+begin
+  refine h.induction _ _;
+  intros,
+  { exact is_atomic_bd_equal _ _ },
+  { exact is_atomic_rel _ _ }
+end
+
+/-- A quantifier-free formula is a formula defined without quantifiers. These are all equivalent
+to boolean combinations of atomic formulas. -/
+inductive is_qf : L.bounded_formula α n → Prop
+| falsum : is_qf falsum
+| of_is_atomic {φ} (h : is_atomic φ) : is_qf φ
+| imp {φ₁ φ₂} (h₁ : is_qf φ₁) (h₂ : is_qf φ₂) : is_qf (φ₁.imp φ₂)
+
+lemma is_qf.induction {P : L.bounded_formula α n → Prop} {φ : L.bounded_formula α n}
```

Comment:
superfluous?

Reply:
As with the other induction lemmas, there appears to be something about `rec_on` that doesn't work right for all these term-mode proofs, but would work if I did it in tactic mode with `exact`.